csrc/abstract_tensor_schedule.cpp

```diff
@@ -0,0 +1,474 @@
+          // the current ValGroup.
+          vg_computed = true;
+          break;
+        }
+        if (!unprocessed_producer_groups.empty()) {
+          // Do not look at other ExprGroups before we try computing these
+          // unprocessed producer groups
+          // At this point vg_computed=false so after this break we will push
+          // the elements of unprocessed_producer_groups onto the stack so that
+          // we process them before proceeding to try the current ValGroup
+          // again.
+          break;
+        }
+      }
+      if (vg_computed) {
+        // Move on to the next ValGroup on the stack
+        continue;
+      }
+
+      if (unprocessed_producer_groups.empty()) {
+        // We did not compute a new expression yet there are no producer groups
+        // left to process. This means all of the defining expressions are
+        // proven uncomputable, so mark vg uncomputable.
+        uncomputable_groups.insert(vg);
+      } else {
+        // There are some unprocessed producer groups that might be computable,
+        // so try again after processing those producer groups
+        eval_stack.push(vg);
+        for (const ValGroup& next_vg : unprocessed_producer_groups) {
+          eval_stack.push(next_vg);
+        }
+      }
+    }
+
+    // Now check whether there is a concrete entry for abs_id (i.e. g). If not,
+    // that means this dimension is missing and should not be included in the
+    // loop domain, so return nullptr.
+    auto it = computed_ids.find(g);
+    return it == computed_ids.end() ? nullptr : it->second;
+  }
+
+ private:
+  const AbstractTensor& abstract_tensor_;
+  ValGraph* graph_;
+
+  std::unordered_set<ValGroup> scheduled_val_groups_;
+};
+
+// TODO: this could potentially belong in id_model/schedule.h
+std::vector<ValGroup> replayExprGroup(
+    ValGraph* graph,
+    ExprGroup eg,
+    const std::vector<ValGroup>& input_groups) {
+  Expr* expr = eg->front();
+  if (auto* m = dynamic_cast<Merge*>(expr)) {
+    NVF_ERROR(input_groups.size() == 2);
+    return {merge(graph, input_groups[0], input_groups[1])};
+  } else if (auto* s = dynamic_cast<Split*>(expr)) {
+    NVF_ERROR(input_groups.size() == 1);
+    auto [outer, inner] =
+        split(graph, input_groups.front(), s->factor(), s->innerSplit());
+    return {outer, inner};
+  } else if (auto* s = dynamic_cast<Swizzle*>(expr)) {
+    NVF_ERROR(input_groups.size() == 2);
+    auto [outer, inner] =
+        swizzle(graph, s->swizzleType(), input_groups[0], input_groups[1]);
+    return {outer, inner};
+  } else {
+    NVF_ERROR(false, "Unhandled scheduling expression ", expr->toString());
+  }
+  return {};
+}
+
+} // namespace
+
+void applyAbstractTransforms(
+    const AbstractTensor& abstract_tensor,
+    const std::vector<TensorView*>& tvs,
+    ValGraph* graph) {
+  AbstractTensorSchedule::apply(abstract_tensor, tvs, graph);
+}
+
+AbstractTensor forwardAroundMissingAxes(
```

Comment:
Something still feels off to me. We are using the Permissive graph here, so if we assume that, the graph should already knows about the broadcast forwarding. Here, it seems we are forwarding whatever missing domains. I'm trying to understand why this is necessary. Could you give me a concrete example?

Reply:
I'll add one as a comment, but the problem is not at the site of the broadcast but before it, where the broadcast axis does not yet exist.

Reply:
Actually, ignore the above comment. The new test seems like a good example.